### PR TITLE
E0608

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of #42229.

cc @Susurrus

### CODE REVIEW

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -4095,6 +4095,27 @@ assert_eq!(!Question::No, true);
 ```
 "##,
 
+E0608: r##"
+An attempt to get an index from a type which doesn't implement the `Index` was
```
Comment (Susurrus) [inline src/librustc_typeck/diagnostics.rs]: This isn't really getting an index, at least in the eyes of the user, is it? It's more trying to index into the type. Maybe rephrase this as "An attempt to index into a type which doesn't implement the `Index` trait was performed." Maybe it actually makes sense to use `std::ops::Index` instead of `Index` as well.

File: src/librustc_typeck/check/mod.rs [hunk 2]
```diff
@@ -3858,13 +3858,9 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
                       None => {
-                          let mut err = self.type_error_struct(
-                              expr.span,
-                              |actual| {
-                                  format!("cannot index a value of type `{}`",
-                                          actual)
-                              },
-                              base_t);
+                          let mut err = struct_span_err!(tcx.sess, expr.span, E0608,
+                                                         "cannot index a value of type `{}`",
```
Comment (Susurrus) [inline src/librustc_typeck/check/mod.rs]: I would use "cannot index into a value of type `{}`" instead

**Review by Susurrus — CHANGES_REQUESTED** [review]

File: src/librustc_typeck/check/mod.rs (hunk 2 above)
Comment (GuillaumeGomez) [inline src/librustc_typeck/check/mod.rs]: That's not the into the problem, that's the index. I think switching the message to this would make the error lose its meaning.

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/librustc_typeck/check/mod.rs (hunk 2 above)
Comment (Susurrus) [inline src/librustc_typeck/check/mod.rs]: This is separate than the `Into` trait stuff. I've always heard indexing described as being into something, like you "index into an array", you don't "index an array". The latter sounds weird to me. With the current PR you use "index into" in the detailed explanation, but the short error title is still "index a value". This should at least be consistent.

**Review by Susurrus — CHANGES_REQUESTED** [review]

File: src/librustc_typeck/check/mod.rs (hunk 2 above)
Comment (GuillaumeGomez) [inline src/librustc_typeck/check/mod.rs]: Ok, let me rewrite this then.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by GuillaumeGomez** [comment]
Updated as well. cc @Susurrus

File: src/librustc_typeck/diagnostics.rs [hunk 3]
```diff
@@ -4095,6 +4095,27 @@ assert_eq!(!Question::No, true);
...
+An attempt to index into a type which doesn't implement the `std::ops::Index`
+trait was performed.
+
+Erroneous code example:
+
+```compile_fail,E0608
+0u8[2]; // error: cannot index into a value of type `u8`
+```
+
+To be able to index a value from a type, it needs to implement the
```
Comment (Susurrus) [inline src/librustc_typeck/diagnostics.rs]: Instead "To be able to index into a type it needs to implement ..."

**Review by Susurrus — CHANGES_REQUESTED** [review]

**Comment by GuillaumeGomez** [comment]
And updated.

cc @frewsxcv @QuietMisdreavus

**Comment by QuietMisdreavus** [comment]
Looks good!

@bors r+
